index.ts

```diff
@@ -1444,6 +1555,146 @@ function findModulesRoot(modulesPaths: string[], userProvidedModules: boolean) {
   );
 }
 
+/*
+Convex.db = DatabaseFake
+Convex.modules = moduleCache
+Convex.syscall = syscallImpl
+Convex.asyncSyscall = asyncSyscallImpl
+Convex.jsSyscall = jsSyscallImpl
+
+
+transaction commit on root only, resetWrites on all bc subtransactions
+
+Convex.functionStack = [{
+  component: "",
+  udfPath: "messages:list"
+}
+]
+Convex.components = {
+  "": {
+  },
+  "aggregate": {
+    components: 
+  }
+  "aggregate/ratelimiter"
+}
+
+t.registerComponent("aggregate", schema, glob for files with functions)
+t.registerComponentFake("aggregate", AggregateFake)
+
+subtransaction -- two functions where second errors (and is caught), we don't commit either because we reset writes
+
+*/
+
+type ComponentInfo = {
+  db: DatabaseFake;
+  modules: ReturnType<typeof moduleCache>;
+};
+
+type FunctionPath = {
+  // e.g. "aggregate"
+  componentPath: string;
+  // e.g. "messages:list"
+  udfPath: string;
+};
+
+type ConvexGlobal = {
+  components: Record<string, ComponentInfo>;
+  transactionManager: TransactionManager;
+  syscall: (op: string, args: any) => any;
+  asyncSyscall: (op: string, args: any) => Promise<any>;
+  jsSyscall: (op: string, args: any) => Promise<any>;
+};
+
+function getConvexGlobal(): ConvexGlobal {
+  return (global as unknown as { Convex: ConvexGlobal }).Convex;
+}
+
+function setConvexGlobal(convex: ConvexGlobal) {
+  if (getConvexGlobal()) {
+    if (getTransactionManager().isInTransaction()) {
+      throw new Error("test began while previous transaction was still open");
+    }
+  }
+  (global as unknown as { Convex: ConvexGlobal }).Convex = convex;
+}
+
+class TransactionManager {
+  // The DatabaseFake is used in the Convex global,
+  // and so it restricts `convexTest` to run one function
+  // at a time.
+  // We force sequential execution to make sure actions
+  // can run mutations in parallel.
+  private _waitOnCurrentFunction: Promise<void> | null = null;
+  private _markTransactionDone: (() => void) | null = null;
+  public functionStack: FunctionPath[] = [];
+
+  async begin(
+    functionPath: FunctionPath,
+    isNested: boolean,
+  ) {
+    if (!isNested) {
```

Comment:
Can we add comment on how we don't have proper transaction serialization for nested calls?